### PR TITLE
Add optional guaranteedChannel to Channel

### PR DESCRIPTION
**Pull request by andrewgordstewart**
A state channel might be a guarantor channel. In this case, it should be invariant throughout the lifetime of the state channel. This PR
- puts it in the `Channel` interface in typescript
- puts it on the `Commitment` struct in solidity
- pulls it off the `Outcome` struct
- removes it as an argument from some smart contract calls.

The rules are changed accordingly:
i) a guarantor channel must have an empty allocation
ii) a ledger channel cannot

This check was also on the `forceMove` method, and is removed from there.

### CODE REVIEW

File: packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol [hunk 1]
```diff
@@ -183,54 +210,62 @@ contract NitroAdjudicator {
...
+        Outcome memory outcome,
+        address recipient,
+        uint amount
+    ) internal pure returns (Outcome memory) {
         uint256[] memory updatedAllocation = outcome.allocation;
         uint256 reduction = 0;
+        uint remainingAmount = amount;
         for (uint i = 0; i < outcome.destination.length; i++) {
             if (outcome.destination[i] == recipient) {
                 // It is technically allowed for a recipient to be listed in the
```
Comment (lalexgap) [inline packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol]: Outside the changes in this PR but I've created #81 for this, since I saw this comment and got burned before.

File: packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol [hunk 2]
```diff
@@ -183,54 +210,62 @@ contract NitroAdjudicator {
...
 
-    function reduce(Outcome memory outcome, address recipient, uint amount) internal pure returns (Outcome memory) { 
+    function reduce(
+        Outcome memory outcome,
+        address recipient,
+        uint amount
+    ) internal pure returns (Outcome memory) {
         uint256[] memory updatedAllocation = outcome.allocation;
         uint256 reduction = 0;
+        uint remainingAmount = amount;
```
Comment (lalexgap) [inline packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol]: Is there a reason we don't want to modify the function parameter `amount`?

File: packages/fmg-core/src/test/framework/transitions.test.ts [hunk 3]
```diff
@@ -81,14 +83,69 @@ describe('Rules', () => {
...
+        commitmentCount: 0,
+      });
+      toCommitment = createCommitment.preFundSetup({
+        ...fromCommitment,
+        turnNum: 1,
+        commitmentCount: 1,
+      });
+    });
+
+    it('allows a valid transition when the allocations are empty in a guarantor channel', async () => {
```
Comment (lalexgap) [inline packages/fmg-core/src/test/framework/transitions.test.ts]: should we add a test for the inverse of this? (ValidTransition failing when allocation is not empty?)

File: packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol (hunk 2 above)
Comment (andrewgordstewart) [reply packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol]: The solidity linter complained about it. I didn't actually try to figure out why.
